modules/dhcp_isc/leases_file.rb

```diff
@@ -0,0 +1,21 @@
+module Proxy::DHCP
+  module ISC
+    class LeasesFile
+      attr_reader :path, :parser, :fd
+
+      def initialize(a_path, config_and_leases_parser)
+        @path = a_path
+        @parser = config_and_leases_parser
+      end
+
+      def hosts_and_leases
+        @fd = File.open(File.expand_path(path), "r") if fd.nil?
+        parser.parse_config_and_leases_for_records(fd.read)
+      end
+
+      def close
+        fd.close unless fd.nil?
```

Comment:
I think `@fd` also has to be set to nil here, otherwise the old fd is used when `hosts_and_leases` is called again from `leases_recreated`:

<pre>
E, [2016-07-29T10:53:27.693723 #5691] ERROR -- : Error occured when monitoring /tmp/dhcp.leases
D, [2016-07-29T10:53:27.693750 #5691] DEBUG -- : closed stream (IOError)
/home/dcleal/code/foreman/smart-proxy/modules/dhcp_isc/leases_file.rb:13:in `read'
/home/dcleal/code/foreman/smart-proxy/modules/dhcp_isc/leases_file.rb:13:in `hosts_and_leases'
/home/dcleal/code/foreman/smart-proxy/modules/dhcp_isc/isc_state_changes_observer.rb:35:in `block in leases_recreated'
/home/dcleal/code/foreman/smart-proxy/modules/dhcp_common/subnet_service.rb:140:in `block in group_changes'
/home/dcleal/.rvm/rubies/ruby-2.3.0/lib/ruby/2.3.0/monitor.rb:214:in `mon_synchronize'
/home/dcleal/code/foreman/smart-proxy/modules/dhcp_common/subnet_service.rb:140:in `group_changes'
/home/dcleal/code/foreman/smart-proxy/modules/dhcp_isc/isc_state_changes_observer.rb:27:in `leases_recreated'
/home/dcleal/code/foreman/smart-proxy/modules/dhcp_isc/inotify_leases_file_observer.rb:25:in `block (2 levels) in monitor_leases'
/home/dcleal/code/foreman/smart-proxy/modules/dhcp_isc/inotify_leases_file_observer.rb:17:in `each'
/home/dcleal/code/foreman/smart-proxy/modules/dhcp_isc/inotify_leases_file_observer.rb:17:in `block in monitor_leases'
</pre>